### DISCUSSION TITLE
Biker map theme (Mapsforge)

### DISCUSSION THREAD

**Discussion by MotoUKRider** [devemux86/cruiser#656] [Show and tell]
There have been some requests for the "Biker" theme to return, which would be a Mapsforge theme written for Cruiser v5. Some efforts were previously made, and it was decided that the "default" day+night themes needed to reach a point where map/object behaviour was *as good as it can be* before we went in 3 directions.  Probably a wise move!

Some of you prefer the OSM-esq colour scheme and of course Biker was the starting block for my original "MotoRider" theme in the v4 VTM era.  If there is interest, I will replicate the original styling/fonts before placing it in this thread for tests.

As of 25th Feb '25, the map looks like this:

![cruiser-biker-theme-1](https://github.com/user-attachments/assets/a21256aa-63a3-45a2-80cd-ba517cce9921)

**Version 51, 28th May**
Minor changes to `islands|islets` and `drain|lake` texts.
[biker-mapsforge.zip](https://github.com/user-attachments/files/20480492/biker-mapsforge.zip)

If you need to know how to display an external theme, [please see this post](https://github.com/devemux86/cruiser/discussions/612#discussion-7028760).

**Comment by devemux86** [comment]
@MotoUKRider Thanks for the suggestion.

It would be a copy of the Motorider xml file with only different colors, right?
It must be easy to maintain, as the same structure changes must be made in all themes.

I am still not sure if we need another theme in Cruiser?
As it complicates things and maintenance. :slightly_smiling_face:

**Comment by DolomitesTom** [comment]
The left picture is better anyway, better contrast.

**Reply by WolfgangZ42** [reply]
The current theme is perfectly adequate for me personally. Based on the above comparison, I don't think the increased maintenance costs are justified. Just my 2 cents.

**Comment by devemux86** [comment]
@MotoUKRider Also I suppose you mean a light theme?
Because another dark theme would have similar colors?

**Comment by bios4** [comment]
While I really liked the Biker theme in Cruiser 4 (still use it on my productive smartphone on the bike) and would appreciate it a lot if it was available for Cruiser 5 again, I understand that it should be well considered in terms of (maintenance) effort.

**Comment by MotoUKRider** [comment]
...and we're back.

Was hoping for a few more users from the woodwork, but it looks like 2 against, and 1 for it - so following some private feedback from @devemux86 the Biker Theme ZIP is in the opening post at the top of this thread.

The basic changes are colour tweaks to the land+roads (more OSM-like) with simple road labelling (no fancy outlines). The rest of the theme, in terms of "rules" and Layers works exactly the same as Cruiser's "default".  Therefore, the larger City/Town/Village fonts are *not* replicated. Of course, if you prefer the white background then you can change the `#e8eae1` colour entries near line:312.

Incidentally, I use [Cx File Explorer](https://play.google.com/store/apps/details?id=com.cxinventor.file.explorer&hl=en_GB) to copy files between Desktop and Phone (Samsung S20) - it can connect to a Windoze shared drive and even has a "recent" list so navigating to a location 83 folders deep is a single button press :clown_face:  The USB-C socket gets a break too!

**Reply by devemux86** [reply]
**KDE Connect** for communication between all devices: https://kdeconnect.kde.org

**Reply by devemux86** [reply]
@MotoUKRider Thanks very nice clean work! :+1:

**Reply by bios4** [reply]
Thanks @MotoUKRider for providing the theme also for Cruiser 5! 👍 😍

Unfortunately my riding season ended yesterday by depositing my license plates and the bike will be picked up and transported to winter storage tomorrow... 😞 

>KDE Connect for communication between all devices

For managing and doing file transfers on Android (and Huawei) devices I can also definitely recommend [MyPhoneExplorer](https://www.fjsoft.at/en/).
Works like a charm for me for almost 13 years now...

**Reply by umpfmpf** [reply]
Also for me it looks very nice! Good Work!!

But is it possible to repeat the road numbers of motorways and expressways at level 11 and 12 a little less often?

**Reply by umpfmpf** [reply]
![Screenshot_20241030_083714_Cruiser](https://github.com/user-attachments/assets/bf02792e-dd4f-4ef9-97ad-1f9c11c32471)

**Reply by devemux86** [reply]
You don't need to drive to check the theme, just browse the map at different zoom levels.

Comparing the Motorider and Biker themes, only the colors should be different.
The structure and frequency of the labels should be the same.

**Reply by umpfmpf** [reply]
It's true, the markings are the same, they just look different. 
I like the biker theme a little better because it's not that intrusive. 
But the frequency is too high at zoom levels 11 and 12.

**Reply by devemux86** [reply]
@umpfmpf There is the layer "More road details" in "Map layers". Do you have it enabled?

**Reply by umpfmpf** [reply]
With the Layer on it looks like this:
![Screenshot_20241030_095515_Cruiser](https://github.com/user-attachments/assets/222d75e7-71fb-4b4a-aece-2541f4708b4f)

**Reply by bios4** [reply]
> You don't need to drive to check the theme, just browse the map at different zoom levels.

Sure you don't NEED to drive, but - for me - only "in theory"...

BECAUSE: only when out on the road and testing in "real driving conditions" one could finally decide about amount of details shown, colors (think about direct sunlight on the screen), etc.

It is a totally different use case when sitting on the sofa and playing with the app/map.
Also, when using it in the car it is different, because you have a different kind of focus on the bike.

At least IMHO.

**Reply by bios4** [reply]
Will the Biker map theme be bundled (in addition to the Motorider themes) with Cruiser 5 in a future version?

**Reply by devemux86** [reply]
> Will the Biker map theme be bundled (in addition to the Motorider themes) with Cruiser 5 in a future version?

Please see my comment above:
- https://github.com/devemux86/cruiser/discussions/656#discussioncomment-11059469

Maintaining themes is not an easy task. We need to think carefully if we need another theme.
- It will also complicate things with https://github.com/devemux86/cruiser/discussions/126

**Reply by devemux86** [reply]
It is better to have 1 very good map theme well maintained,
than 2-3 map themes that are not very well maintained.

**Reply by MotoUKRider** [reply]
@umpfmpf Sure, just reduce the font-size in your settings to the normal 100% setting. I think we had this before... and my reply was along the lines of: *"You can't increase the font-size and then complain that the fonts are too big."*

There's no easy way to accommodate "accessibility" (or visual enhancement) in this way without directly editing the theme, unless @devemux86 knows better. There are settings to reduce the frequency/population of labels, and it took some time to find the best approach (and they vary between Desktop and App) via trial'n'error.  I think the only way, for now, to cope with a big font size on the map is for you to edit the XML and find your own `repeat-start` and `repeat-gap` values that work with the visual configuration you have.

**Reply by devemux86** [reply]
@MotoUKRider 

Alternatively, we can change several street labels from captions to path texts (in small zooms),
such as regular street names. But then these ref names will be much less frequent on the map.

**Reply by umpfmpf** [reply]
The problem is not the font sizeor that it is too big. I only need it to drive, otherwise I can't read anything. 
The problem is the number of repetitions on the same street in ZoomLevel 11 and 12.

**Reply by MotoUKRider** [reply]
@devemux86 I don't think there's a "best fit" solution that pleases @umpfmpf or any other possible visual requirement.

The whole lot would have to be in several Layers, ie: "Normal View", "Large Font", "Not-too-big-but-enough-so-I-can-see-it-with-my-spare-glasses Font" etc. Each of those would have their own `repeat` parameters, not-to-mention `priority` which becomes a thing when POIs (like Fuel) become visible... that would bring us into the *"I want to see fuel icons except when there's a motorway label but when there's parking available I don't want to see cafes or churches unless I'm on a Secondary road and it's a Sunday"* situation again! :confused: 

Another case of change/fix 1 thing and 2 other aspects break :warning:  I would suggest that @umpfmpf modifies the theme to their tastes by adjusting the parameters I mentioned because the whole theme cannot be scaled in terms of % by an App Setting. This is why the theme XML is here: You can experiment with the values for your own custom situation.

**Reply by devemux86** [reply]
This is why I still find the "More road details" layer confusing and prefer not to have it.

We need to have a theme with the best global sane appearance.
Adjustments can be made on top with the Cruiser map settings.
Or advanced users can modify the theme xml.

My remark was that by changing some road captions to path texts, we can have less ref labels filling the map.

**Reply by MotoUKRider** [reply]
@devemux86 The rationale behind "More road details" was allowing you to follow the road labels when planning, and offers the option of seeing them without having to zoom in+out too much.

As evidence, I offer GoogleMaps vs 2009 - It was better back in my day!

![google-maps-2009vs2024](https://github.com/user-attachments/assets/54494503-b851-4ee8-85f1-21c14af06409)

**Reply by devemux86** [reply]
@MotoUKRider Do you mean you prefer the left image?

A middle view would be better, as the left image is very empty of road information,
while the right image is too full of road information that steals attention away from the cities.

**Reply by MotoUKRider** [reply]
@devemux86 No, I think it's overly green and terribly basic. It looks like they've even put chlorine in the north sea, too.

That said, these days it's not really a map for looking at in terms of pretty colours - it's for looking-up a business and doing that tediously efficient PointA-to-PointB navigation that normal people like :rofl:

**Reply by waterhead** [reply]
Yes, the current Google theme is simply horrid. Was definitely better in the past!

**Comment by devemux86** [comment]
@MotoUKRider 

You can add the Chinese layer names from the [Motorider theme](https://github.com/mapsforge/mapsforge/tree/master/mapsforge-themes/src/main/resources/assets/mapsforge).

**Reply by MotoUKRider** [reply]
@devemux86 Thanks - now updated the ZIP in the top post :wrench:

**Comment by MotoUKRider** [comment]
Revision 8 now updated behaviour (`construction` roads) to match main theme.

**Comment by devemux86** [comment]
@MotoUKRider Do you use the OpenStreetMap website colors in the Biker theme?

**Reply by MotoUKRider** [reply]
@devemux86 Not like-for-like, but similar (and more by accident than design).  The obvious exception is `tertiary` roads having a slight pink fill in this theme.  These were, I believe, the original theme colours which I simply replicated and I don't think any "core" element was changed.

So, essentially, it's the "motorider" XML (to keep the formatting/zoom rules) with the "biker" colours copied from the VTM version :art:

**Reply by devemux86** [reply]
@MotoUKRider The old Biker theme in VTM is not perfect, as we didn't update it. :slightly_smiling_face: 

If OpenStreetMap colors are already used, we could use all OpenStreetMap colors for a similar appearance.
OpenStreetMap is used in many websites by millions of users and they are already used to these colors.

**Comment by devemux86** [comment]
**Poll**: Do you all think the Biker theme should be included in Cruiser?
Do you use it?

We will need some improvements before that:
- **Colored** road `ref` labels for better visibility (like Motorider)

- **Dark** version, so it can be switched in navigation

**Reply by bios4** [reply]
> Do you all think the Biker theme should be included in Cruiser?

I think it would be a nice addition, yes...

> Do you use it?

Yes! But not "as is" but tweaked to my likings (i.e. a combination of Biker / Motorider light).

So for me personally there is no preference for "yes" or "no" if it should be included.
As long as it is available to be activated separately, I'm fine. 😉

**Reply by WolfgangZ42** [reply]
For me it's not necessary.

**Comment by DolomitesTom** [comment]
I would like to have an alternative in the Biker theme.

**Comment by MotoUKRider** [comment]
Revision 12 now added, a couple of minor changes like road-ref labels [as per this post](https://github.com/devemux86/cruiser/discussions/656#discussioncomment-12091608). A couple of mis-aligned `stroke-width` values also fixed.

Here is a snippet from "Desktop" to show the change:
![cruiser-biker-25feb](https://github.com/user-attachments/assets/5fcfa00a-6be3-48d7-bc13-f4015873d0c1)

If you want to compare with earlier, the preview images at the top of this thread have the original "black" font.

**Reply by Ciutnik** [reply]
I am sorry, but it seems to me that (perhaps) it would be better if the colours of the labels were related to the colours of the roads they refer to. Especially at a junction, it is difficult to match the label to the road. But this is rather a minor thing 🙂

**Reply by devemux86** [reply]
@MotoUKRider @Ciutnik 

Indeed, the `ref` labels better use the theme's road colors.

**Reply by MotoUKRider** [reply]
@Ciutnik @devemux86  Here are the road-ref labels using their respective colours, standard+inverted:

![cruiser-biker-roadrefs](https://github.com/user-attachments/assets/eb48146e-373a-4373-942d-afd8c96b8e08)

It seems less readable, but I don't use the theme, so I'm not that fussed :smile:

**Reply by devemux86** [reply]
@MotoUKRider 

Like in 2nd screenshot, but with **black** text?

**Reply by Ciutnik** [reply]
I didn't mean identical colors, but related, similar to the colors of the roads, but with more contrast...

**Reply by MotoUKRider** [reply]
I'm not sure if black text is better with the road `stroke` color?

![cruiser-biker-roadrefs-2](https://github.com/user-attachments/assets/f521a374-07f4-467e-a4d0-16e58aee9795)

In the original screenshots at the top, the `stroke` (pen color) was white, so offset from the surrounding colors (road/land etc). When it's the same as the road it doesn't "stand out" as much, in my opinion.  So, for contrast, the original feels best - possibly with a larger `stroke` to create a more visible border?

Here is stroke=6 (up from 4) with white and off-white (the actual map background) colors:
![cruiser-biker-roadrefs-stroke6](https://github.com/user-attachments/assets/11e60aab-9e56-4763-9029-58727ed287ea)

For those not familiar with the styling, we use `stroke` as a "border" because we do not have a "box" object in this type of theme.
OpenStreetMap (which Biker is reasonably faithful to) uses a box with a slightly lighter fill than the road, with black text.

**Reply by devemux86** [reply]
@MotoUKRider 

> OpenStreetMap (which Biker is reasonably faithful to) uses a box with a slightly lighter fill than the road, with black text.

That's why I made the above suggestion, but it may indeed not be very readable.

> So, for contrast, the original feels best - possibly with a larger stroke to create a more visible border?

> Here is stroke=6 (up from 4) with white and off-white (the actual map background) colors:

This seems better.
I don't have strong opinion about the stroke width or how much "white". :slightly_smiling_face:

**Comment by MotoUKRider** [comment]
Lunchtime crayons...

![cruiser-biker-roadrefs-osm](https://github.com/user-attachments/assets/3f668f36-465d-4c07-9577-8ba465b99919)

**Comment by MotoUKRider** [comment]
I've re-upped the Biker ZIP theme at the top of this thread - and updated the in-app screenshots.

Here's a close-up of the new labels:
![cruiser-biker-road-labels](https://github.com/user-attachments/assets/99f8940e-1a50-4740-9ad9-59b8672fd6d7)

It's probably worth mentioning that this now takes this and our standard "Light" theme *out* of alignment in terms of XML like-for-like content as it's not just the colors that have been changed, but also `stroke`. Older rules remain commented-out but the XMLs will be matched once we all happy with these changes.

For all themes, there will probably be some tweaks to `pathtext` labels (that is, a label which follows the curvature of the road/river) - I'm not sure a big splodge of `stroke` is ideal for close-in road names. The main concern is styling the map and its labels for planning *and* navigating - you obviously want a label you can read but not something that covers other things too much.

**Reply by devemux86** [reply]
@MotoUKRider Have you tried semi-transparency in the stroke color?

**Reply by MotoUKRider** [reply]
@devemux86 Not yet, I'll leave this one to stew for a while before making any further changes.

**Reply by devemux86** [reply]
@MotoUKRider 

> this now takes this and our standard "Light" theme out of alignment in terms of XML like-for-like content as it's not just the colors that have been changed, but also stroke.

You mean that the lines remain the same, but we have changes in both `fill` and `stroke` parameters?
This is not a problem, as long as the structure remains the same.

> Older rules remain commented-out but the XMLs will be matched once we all happy with these changes.

Please do this at the end, so that the structure of all themes and their lines are the same for easy comparison!

**Reply by devemux86** [reply]
@MotoUKRider 

Have you changed in some labels the `font-style` from `bold` to `normal`?
With `normal` they are less readable?

**Reply by MotoUKRider** [reply]
@devemux86 Oh, yes - they looked a bit **smudgy** in bold. It helps to increase the font-size but then we start going around in circles. I'll double-check both `normal` and `bold` in App but my first attempt above was to mimic the OSM style.

**Comment by MotoUKRider** [comment]
Revision 13 now available - fixed a railway-bridge issue.

**Comment by devemux86** [comment]
Biker map theme will be available in the next Cruiser version.

**Comment by devemux86** [comment]
New Biker map theme for Mapsforge is available in:
- https://github.com/devemux86/cruiser/discussions/734

**Comment by MotoUKRider** [comment]
Revision 14 now updated to include "Vegetation" Layer (enabled by default for normal map view) - turn it off to hide all natural "green" land features like forest, meadow, grass etc.

Some POI caption fixes and minor edits.

**Comment by devemux86** [comment]
- Biker map theme (Mapsforge) v14 is available in https://github.com/devemux86/cruiser/discussions/757

**Comment by MotoUKRider** [comment]
Version 48 now available in the top post.  This has the revised `"More map details"` Layer (replacing "More road details") with additional detail for cycleways/paths etc.  This brings the themes running under the "Mapsforge" engine in-line with the 3D OpenGL/VTM ones.

**Comment by MotoUKRider** [comment]
v49 now available in the top post which adds "Show toll roads" Layer. This will highlight roads that have `toll=yes`. Icons for `toll_booth` have also been added which were previously missing.

**Comment by MotoUKRider** [comment]
v50 now available which has a newly renamed "Toll roads" Layer (was "Show toll roads").

**Comment by devemux86** [comment]
Mapsforge map themes will all be maintained here:
- https://github.com/devemux86/cruiser/discussions/823
